### PR TITLE
Avoid using cached bindings when the module isn't loaded from cache

### LINKED ISSUE #8421: Deleting a cache file causes run-time compiler error to be thrown
The repro that seems to work for me is:
1. first do a clean `buildEngineDistribution`.
2. then, remove `rm .\built-distribution\enso-engine-0.0.0-dev-windows-amd64\enso-0.0.0-dev\lib\Standard\Database\0.0.0-dev\.enso\cache\ir\0.0.0-dev\Standard\Database\Connection\SQLite_Format.ir` (the path may differ slightly on other platforms)
3. and run `enso run .\test\Table_Tests\src\IO\Formats_Spec.enso`

For me it fails with
```
    - [FAILED] should fail gracefully when provided with an unsupported format [50ms]
        Reason: An unexpected panic was thrown: (Compile_Error.Error 'The name `Nothing` could not be found')
        at <enso> case_branch(C:\NBO\enso\built-distribution\enso-engine-0.0.0-dev-windows-amd64\enso-0.0.0-dev\lib\Standard\Database\0.0.0-dev\src\Connection\SQLite_Format.enso:33:18-24)
        at <enso> SQLite_Format.type.for_file_write(C:\NBO\enso\built-distribution\enso-engine-0.0.0-dev-windows-amd64\enso-0.0.0-dev\lib\Standard\Database\0.0.0-dev\src\Connection\SQLite_Format.enso:30-33)
        at <enso> <anonymous>(C:\NBO\enso\built-distribution\enso-engine-0.0.0-dev-windows-amd64\enso-0.0.0-dev\lib\Standard\Base\0.0.0-dev\src\System\File_Format.enso:69:24-44)
        at <enso> reader<arg-2>(C:\NBO\enso\built-distribution\enso-engine-0.0.0-dev-windows-amd64\enso-0.0.0-dev\lib\Standard\Base\0.0.0-dev\src\System\File_Format.enso:39:22-44)
        at <enso> reader(C:\NBO\enso\built-distribution\enso-eng
```

### PR DESCRIPTION
**Pull request by JaroslavTulach**
### Pull Request Description

Fixes #8421.

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      style guides.
- All code has been tested:
  - [x] Manually tested.

### CODE REVIEW

**Comment by JaroslavTulach** [comment]
I have manually verified that the [failing reproducer](https://github.com/enso-org/enso/issues/8421#issuecomment-1876296226) now works:
```
enso$ rm -rf *; git checkout -f .
enso$ git log | head -n5
commit a1207e029d7780a54a6d09ca2665f7370493b4f8
Author: Radosław Waśko <radoslaw.wasko@enso.org>
Date:   Thu Jan 4 04:57:05 2024 +0100

    Unify `File_Format_Metadata` with `File_For_Read` (#8628)
enso$ sbt --java-home /graalvm buildEngineDistribution
enso$ rm built-distribution/enso-engine-0.0.0-dev-linux-amd64/enso-0.0.0-dev/lib/Standard/Database/0.0.0-dev/.enso/cache/ir/0.0.0-dev/Standard/Database/Connection/SQLite_Format.ir
$ rm -rf ~/.local/share/enso/cache/ir/
enso$ ./built-distribution/enso-engine-0.0.0-dev-linux-amd64/enso-0.0.0-dev/bin/enso --in-project test/Table_Tests/ --run test/Table_Tests/src/IO/Formats_Spec.enso 
[WARN] [2024-01-04T06:40:00+01:00] [org.enso.librarymanager.local.DefaultLocalLibraryProvider] Local library search path [***/libraries] does not exist.
[WARN] [2024-01-04T06:40:03+01:00] [enso.org.enso.interpreter.caches.ModuleCache] Unable to load a cache [Standard.Database.Connection.SQLite_Format]: enso/built-distribution/enso-engine-0.0.0-dev-linux-amd64/enso-0.0.0-dev/lib/Standard/Database/0.0.0-dev/.enso/cache/ir/0.0.0-dev/Standard/Database/Connection/SQLite_Format.ir
Various File Format support on Table:  [6/6, 1549ms]
    - should be able to be written as CSV, Excel [623ms]
    - should be able to be written as JSON using Table.write [423ms]
    - should write JSON tables [254ms]
    - should append to JSON tables [177ms]
    - should fail to append to JSON non-arrays [32ms]
    - should fail gracefully when provided with an unsupported format [36ms]
Writing to csv files:  [3/3, 620ms]
    - should write to a temporary csv file part of the data if context is disabled [270ms]
    - should create a backup file if overwriting [118ms]
    - should support appending [232ms]
Writing to json files:  [3/3, 992ms]
    - should write to a temporary json file part of the data if context is disabled [920ms]
    - should create a backup file if overwriting [33ms]
    - should support appending [37ms]
12 tests succeeded.
0 tests failed.
0 tests skipped.
0 groups of tests skipped.
```

File: engine/runtime/src/main/java/org/enso/interpreter/caches/Cache.java [hunk 1]
```diff
@@ -218,6 +218,11 @@ public final Optional<T> load(EnsoContext context) {
                   }
 
                   logger.log(logLevel, "Unable to load a cache [" + logName + "]");
+                } catch (IOException e) {
+                  logger.log(
```
Comment (hubertp) [inline engine/runtime/src/main/java/org/enso/interpreter/caches/Cache.java:222]: Twice the same log?

Comment (JaroslavTulach) [reply engine/runtime/src/main/java/org/enso/interpreter/caches/Cache.java:222]: My goal was to log just the message with the default log level, but allow logging of the whole stacktrace when somebody increases the log level. What's the right way to do it?

Comment (hubertp) [reply engine/runtime/src/main/java/org/enso/interpreter/caches/Cache.java:222]: I would say that whether to include or not the stacktrace is the role of the logging configuration.
See the difference in `%nopex` in https://github.com/enso-org/enso/blob/develop/engine/launcher/src/main/resources/application.conf#L30 configuration - no stacktraces in console but full stacktrace in the log file.

Comment (JaroslavTulach) [reply engine/runtime/src/main/java/org/enso/interpreter/caches/Cache.java:222]: Is this what you want me to do? 86ebf06

Comment (JaroslavTulach) [reply engine/runtime/src/main/java/org/enso/interpreter/caches/Cache.java:222]: > no stacktraces in console but full stacktrace in the log file.

I believe the https://github.com/enso-org/enso/issues/8421#issuecomment-1876296226 says the stacktrace was visible in the console as well.

**Review by Akirathan — APPROVED** [review]

**Comment by radeusgd** [comment]
Thanks so much for fixing this! 🎉

**Pull request merged**